src/Aspire.Hosting/Dcp/KubernetesService.cs

```diff
@@ -280,16 +285,78 @@ public void Dispose()
 
     private static bool IsRetryable(Exception ex) => ex is HttpRequestException || ex is KubeConfigException;
 
-    private void EnsureKubernetes()
+    private readonly SemaphoreSlim _kubeconfigReadSemaphore = new(1);
+
+    private async Task EnsureKubernetesAsync(CancellationToken cancellationToken = default)
     {
-        if (_kubernetes != null) { return; }
+        // Return early before waiting for the semaphore if we can.
+        if (_kubernetes != null)
+        {
+            logger.LogDebug("Kubernetes ensured at first chance shortcut.");
+            return;
+        }
+
+        while (!await _kubeconfigReadSemaphore.WaitAsync(100, cancellationToken).ConfigureAwait(false))
+        {
+            logger.LogDebug("Waiting for semaphore to read kubeconfig.");
+        }
+
+        // Return late after getting the semaphore but also probably after
+        // another thread has gotten the config loaded.
 
-        lock (Model.Dcp.Schema)
+        if (_kubernetes != null)
         {
-            if (_kubernetes != null) { return; }
+            logger.LogDebug("Kubernetes ensured at second chance shortcut.");
+            _kubeconfigReadSemaphore.Release();
+            return;
+        }
 
-            var config = KubernetesClientConfiguration.BuildConfigFromConfigFile(kubeconfigPath: locations.DcpKubeconfigPath, useRelativePaths: false);
-            _kubernetes = new DcpKubernetesClient(config);
+        try
+        {
+            // This retry was created in relation to this comment in GitHub:
+            //
+            //     https://github.com/dotnet/aspire/issues/2422#issuecomment-2016701083
+            //
+            // It looks like it is possible for us to attempt to read the file before it is written/finished
```

Comment:
In Aspire P5 this won't be true anymore.

Reply:
Tidied this comment up.